src/state/router.js

```diff
@@ -0,0 +1,142 @@
+		method: encodeString( request.method ),
+		pathValues: encodeObject( request.pathValues ),
+		queryParams: encodeObject( request.queryParams ),
+		url: encodeString( request.url ),
+		api: encodeString( ui.api ),
+		version: encodeString( ui.version ),
+	};
+
+	// Discard empty params
+	Object.keys( params ).forEach( key => ! params[ key ] && ( delete params[ key ] ) );
+
+	// Construct url
+	const urlParams = new URLSearchParams( params );
+	const url = window.location.origin + window.location.pathname + '?' + urlParams.toString();
+
+	return url;
+};
+
+export const getStateFromUrl = () => {
+	const urlParams = getUrlParams();
+
+	if ( urlParams.toString() === '' ) {
+		return false;
+	}
+
+	const state = { request: { ...defaultRequestState }, ui: { ...defaultUiState } };
+
+	try {
+		for ( const [ key, value ] of urlParams.entries() ) {
+			switch ( key ) {
+				case 'bodyParams':
+					state.request.bodyParams = decodeObject( value );
+					break;
+				case 'method':
+					state.request.method = decodeString( value );
+					break;
+				case 'pathValues':
+					state.request.pathValues = decodeObject( value );
+					break;
+				case 'queryParams':
+					state.request.queryParams = decodeObject( value );
+					break;
+				case 'url':
+					state.request.url = decodeString( value );
+					break;
+				case 'api':
+					state.ui.api = decodeString( value );
+					break;
+				case 'version':
+					state.ui.version = decodeString( value );
+					break;
+				default:
+					break;
+			}
+		}
+	} catch ( e ) {
+		// Fail without breaking the app
+		console.error( 'Could not parse state from url params.', e );
+		return false;
+	}
+
+	return state;
+};
+
+const router = ( { getState, dispatch } ) => {
+
+	// Determine if we need to load and select the endpoint
+	let isInitializingEndpoint = false;
+
+	const { ui = {} } = getStateFromUrl();
+	const endpointUrlParam = getUrlParams().get( 'endpoint' );
+	const endpointPathLabeled = endpointUrlParam ? decodeURIComponent( endpointUrlParam ) : false;
+	const apiName = ui.api;
+	const apiVersion = ui.version;
+
+	if ( endpointPathLabeled && apiName && apiVersion ) {
+		loadEndpoints( apiName, apiVersion )( dispatch );
+		isInitializingEndpoint = true;
+	}
+
+	return next => action => {
+		const state = getState();
+
+		switch ( action.type ) {
+			case REQUEST_TRIGGER:
+				const url = getUrlFromState( state );
+				window.history.pushState( {}, document.title, url );
+				break;
+			case API_VERSIONS_RECEIVE:
+				// Once the endpoint is loaded, select the endpoint.
+				if ( isInitializingEndpoint ) {
+					const endpoints = getEndpoints( state, apiName, apiVersion );
+					const endpoint = endpoints.find( ( { pathLabeled } ) => pathLabeled === endpointPathLabeled );
+					if ( endpoint ) {
+						dispatch( { type: REQUEST_SELECT_ENDPOINT, payload: { endpoint } } );
+					}
+					isInitializingEndpoint = false;
+				}
+				break;
```

Comment:
nice queueing of the initial request.
do we need to make allotments for logged-out views here or will this only run if someone is logged in?